test/battle/move.c

```diff
@@ -82,55 +82,34 @@ SINGLE_BATTLE_TEST("Turn order is determined randomly if priority and Speed tie
-    PARAMETRIZE { a = 1; b = 2; c = 0; d = 3; }
-    PARAMETRIZE { a = 1; b = 2; c = 3; d = 0; }
-    PARAMETRIZE { a = 1; b = 3; c = 0; d = 2; }
-    PARAMETRIZE { a = 1; b = 3; c = 2; d = 0; }
-    PARAMETRIZE { a = 2; b = 0; c = 1; d = 3; }
-    PARAMETRIZE { a = 2; b = 0; c = 3; d = 1; }
-    PARAMETRIZE { a = 2; b = 1; c = 0; d = 3; }
-    PARAMETRIZE { a = 2; b = 1; c = 3; d = 0; }
-    PARAMETRIZE { a = 2; b = 3; c = 0; d = 1; }
-    PARAMETRIZE { a = 2; b = 3; c = 1; d = 0; }
-    PARAMETRIZE { a = 3; b = 0; c = 1; d = 2; }
-    PARAMETRIZE { a = 3; b = 0; c = 2; d = 1; }
-    PARAMETRIZE { a = 3; b = 1; c = 0; d = 2; }
-    PARAMETRIZE { a = 3; b = 1; c = 2; d = 0; }
-    PARAMETRIZE { a = 3; b = 2; c = 0; d = 1; }
-    PARAMETRIZE { a = 3; b = 2; c = 1; d = 0; }
-
-    order[a] = playerLeft;
-    order[b] = playerRight;
-    order[c] = opponentLeft;
-    order[d] = opponentRight;
+    u16 hpMon1, hpMon2;
 
     PASSES_RANDOMLY(1, 24, RNG_SPEED_TIE);
 
+    ASSUME(gMovesInfo[MOVE_ENDEAVOR].effect == EFFECT_ENDEAVOR);
+    ASSUME(gMovesInfo[MOVE_LIFE_DEW].effect == EFFECT_JUNGLE_HEALING);
+    ASSUME(gMovesInfo[MOVE_CRUSH_GRIP].effect == EFFECT_VARY_POWER_BASED_ON_HP);
+    ASSUME(gMovesInfo[MOVE_SUPER_FANG].effect == EFFECT_SUPER_FANG);
     GIVEN {
-        PLAYER(SPECIES_WOBBUFFET) { Speed(1); }
+        PLAYER(SPECIES_WOBBUFFET) { MaxHP(480); HP(360); Defense(100); Speed(1); }
         PLAYER(SPECIES_WYNAUT) { Speed(1); }
-        OPPONENT(SPECIES_WOBBUFFET) { Speed(1); }
+        OPPONENT(SPECIES_WOBBUFFET) { Attack(100); Speed(1); }
         OPPONENT(SPECIES_WYNAUT) { Speed(1); }
     } WHEN {
-        TURN { MOVE(playerLeft, MOVE_SPLASH); MOVE(playerRight, MOVE_SPLASH); MOVE(opponentLeft, MOVE_SPLASH); MOVE(opponentRight, MOVE_SPLASH); }
+        TURN { MOVE(playerLeft, MOVE_ENDEAVOR, target: opponentLeft); MOVE(playerRight, MOVE_LIFE_DEW); MOVE(opponentLeft, MOVE_CRUSH_GRIP, target: playerLeft, WITH_RNG(RNG_DAMAGE_MODIFIER, 0)); MOVE(opponentRight, MOVE_SUPER_FANG, target: playerLeft); }
     } SCENE {
-        ANIMATION(ANIM_TYPE_MOVE, MOVE_SPLASH, order[0]);
-        ANIMATION(ANIM_TYPE_MOVE, MOVE_SPLASH, order[1]);
-        ANIMATION(ANIM_TYPE_MOVE, MOVE_SPLASH, order[2]);
-        ANIMATION(ANIM_TYPE_MOVE, MOVE_SPLASH, order[3]);
+        HP_BAR(opponentLeft, captureHP: &hpMon2);
+        HP_BAR(playerLeft, captureHP: &hpMon1);
+        HP_BAR(playerLeft, captureHP: &hpMon1);
+        HP_BAR(playerLeft, captureHP: &hpMon1);
+    } THEN {
+        EXPECT_EQ(hpMon1, 202);
+        EXPECT_EQ(hpMon2, 360);
+        //  Any of the following pairs of hp values work:
+        //  (188, 360), (189, 360), (261, 360), (235, 360), (262, 360), (202, 360),
+        //  (189, 378), (189, 189), (189, 480), (138, 480), (188, 240), (188, 188),
+        //  (262, 262), (262, 142), (202, 403), (202, 202), (262, 283), (202, 283),
+        //  (235, 180), (261, 180), (235, 235), (235, 300), (261, 141), (261, 261)
```

Comment:
Do you think it would be possible to record all pairs of HP values, and then check that each of those appears exactly once after all 24 cases have run?

I think as-written this test only verifies that a particular order occurs 1/24 times. I'm not worried that the test is wrong, but maybe in the future somebody will change something about the implementation of speed ties which ends up biasing them again and it would be nice if this test could catch that.

Reply:
It should be doable, it'll probably look a bit ugly though.

Reply:
So it was harder than expected, I need a way to store the result between runs.
Declaring a `u8 occurenceCounts[24] = {...}` at the top of the test doesn't seem to be persistent.

Reply:
Ah, yeah, the function will get reentered a bunch (at least once per "trial", but actually quite a bit more than that). A hacky `static u8 occurrenceCounts[24] = { ... };` might do the trick, at the cost of using up 24 bytes of IWRAM (or EWRAM if you tag it with `EWRAM_DATA`).

I'm not sure if there's an actually good way to make variables persistent otherwise. The `PARAMETRIZE`d tests have a way to carry information between the cases (I think you list those members in the parameters of the `TEST_*` function?), but I think it doesn't work for `PASSES_RANDOMLY` out of the box. Maybe it wouldn't be too hard to add it :thinking:

Reply:
Looking at it quickly it looks like the `PARAMETRIZE`d tests shove the data into `sBackupMapData`.
I tried to declare the array the same way `PARAMETRIZE`d test works, but it didn't work, it said it was undeclared.

Reply:
Ok, that was me forgetting how the `PARAMETRIZE` worked.
`PASSES_RANDOMLY is incompatible with results` is the real error.

Reply:
I made it work.
I had to introduce a new variable in `BattleTestRunnerState` to keep track of the occurrences.
Tested it with changing the speed tie array and it resulted in a test failure.
I guess there is possibly some way that the test could succeed with some combination of results, but it should be extremely unlikely.